### PR TITLE
feat: basic LRU implementation

### PR DESCRIPTION
**Pull request by unw9527**
As title. Unit tests are included.

### CODE REVIEW

File: storage-node/src/cache/lru.rs [hunk 1]
```diff
@@ -1,38 +1,174 @@
...
+        self.cache_map.insert(key.clone(), value);
+        while self.curr_capacity > self.max_capacity {
+            if let Some((key, (_, file_size))) = self.cache_map.pop_front() {
+                println!("-------- Evicting Key: {:?} --------", key);
+                self.curr_capacity -= file_size;
+            }
+        }
+    }
+
+    fn peek<Q>(&mut self, key: &Q) -> Option<&ParpulseCacheValue>
```
Comment (xx01cyx) [inline storage-node/src/cache/lru.rs]: We could use `&self` here if we're not going to modify `self`.

File: storage-node/src/cache/mod.rs [hunk 2]
```diff
@@ -1,22 +1,36 @@
...
-pub trait ParpulseCache {
-    // TODO: Consider making the cache lock-free so that we can have `&self`
-    // instead of `&mut self` here, which enables multiple calls to the cache at
-    // the same time.
-    fn get_value(&mut self, key: &ParpulseCacheKey) -> Option<ParpulseCacheValue>;
-
-    fn set_value(&mut self, key: ParpulseCacheKey, value: ParpulseCacheValue);
+pub trait ParpulseCache
+where
+    ParpulseCacheKey: Eq + Hash,
```
Comment (xx01cyx) [inline storage-node/src/cache/mod.rs]:
```suggestion
```

Comment (unw9527) [reply storage-node/src/cache/mod.rs]: Done

**Comment by xx01cyx** [comment]
Per offline discussion, @lanlou1554 mentions that it's more intuitive to have `ParpulseCache` trait with generic types and have `LruCache` with concrete types. On my second thought, however, I still feel like we should use generic types for `LruCache` and use concrete types for `ParpulseCache` trait. There are several reasons:
1. Implementation-wise, we can view `LruCache` (and also other cache algorithms we'll potentially implement in the future) as a **general** component/tool. There might be other use cases in the system that need `LruCache` (though I can't think of any example for now) -- after all it's a very common and widely adopted concept in all kinds of scenarios.
2. Semantic-wise, `LruCache` is just a wrapper for the least-recent-used logic and should not make any assumption about the key/value type. The `ParpulseCache` trait, however, inherently implies that it's used for the Parpulse I/O service and should thereby qualify the key/value type for the system. We can view the `ParpulseCache` trait as something that is **exclusively** used by our `StorageManager`, which is intended to serve the requests from 721 Execution Engines with a fixed key type. If we plan to extend Parpulse for general usage in the future (e.g. adapting Parpulse for other execution engines), we could do the modifications then. We can just keep the interface non-generic for now.

**Comment by xx01cyx** [comment]
~...but I do think of a case where we should make `ParpulseCache` generic.~ Suppose we want to delegate the operation of "writing back file contents" to the cache, then we need to pass sth like `Bytes` or `BytesStream` to the cache as well when we call `put` on the cache. ~If we'd like both a sync version and an async version then we need distinct value types for each version.~ We can have a `CacheManager` for this which can do both sync and async disk-writes but we don't need to make `ParpulseCache` generic.

File: storage-node/src/cache/lru.rs [hunk 3]
```diff
@@ -1,38 +1,178 @@
...
+    fn clear(&mut self) {
+        self.cache_map.clear();
+        self.curr_capacity = 0;
     }
 }
 
-impl ParpulseCache for LruCache<ParpulseCacheKey, ParpulseCacheValue> {
-    fn get_value(&mut self, key: &ParpulseCacheKey) -> Option<ParpulseCacheValue> {
-        self.get(key)
+mod tests {
```
Comment (xx01cyx) [inline storage-node/src/cache/lru.rs:113]:
```suggestion
#[cfg(test)]
mod tests {
```
This allows the compiler to compile the mod only when testing.

File: storage-node/src/cache/mod.rs [hunk 4]
```diff
@@ -9,14 +20,35 @@ pub type ParpulseCacheKey = String;
...
+impl CacheValue for ParpulseCacheValue {
+    type Value = String;
+
+    fn into_value(self) -> String {
+        self.0
+    }
+
+    fn as_value(&self) -> &String {
+        &self.0
+    }
```
Comment (unw9527) [inline storage-node/src/cache/mod.rs]:
```suggestion
    fn into_value(self) -> Self::Value {
        self.0
    }

    fn as_value(&self) -> &Self::Value {
        &self.0
    }
```

Wouldn't this make more sense? Other than this, LGTM.

File: storage-node/src/cache/mod.rs [hunk 5]
```diff
@@ -9,14 +23,45 @@ pub type ParpulseCacheKey = String;
...
+    fn len(&self) -> usize;
+
+    /// Returns the total size of the objects in the cache.
+    fn size(&self) -> usize;
+
+    fn is_empty(&self) -> bool;
+
+    fn max_capacity(&self) -> usize;
+
+    fn set_max_capacity(&mut self, capacity: usize);
```
Comment (xx01cyx) [inline storage-node/src/cache/mod.rs:64]: Getter & Setter naming convention in Rust: https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis

**Review by xx01cyx — APPROVED** [review]

**Pull request merged**